[CVP-2052] Fix failures when running the midstream-cvp image within an OpenShift pod

samvarankashyap opened the pull request:
the problem arise from the fact that the container runs within an OpenShift pod in more restrictive manner. For e.g., it doesn’t run as root.

For resolving this issue, we have discussed two alternative solutions:

Using product-ci-cd's entrypoint script:https://code.engineering.redhat.com/gerrit/plugins/gitiles/product-ci-cd/+/refs/heads/master/container/midstream/entrypoint.sh and also adding the following to the Dockerfile:

dirgim commented on Dockerfiles/midstream/Dockerfile:
```diff
@@ -13,6 +13,8 @@ RUN export ARCH=$(case $(arch) in x86_64) echo -n amd64 ;; aarch64) echo -n arm6
     mkdir /project/output && \
     dnf install --setopt=install_weak_deps=False -y git-core ansible skopeo && \
     dnf clean all
+ADD ./entrypoint.sh /entrypoint.sh
```
Can you move this `ADD` instruction to be above line 6 and then incorporate the `RUN chmod g+x /entrypoint.sh && chmod g+w /etc/passwd` commands into the `RUN` instruction at line 6?

samvarankashyap replied:
Sure I can do that.

lslebodn commented on Dockerfiles/midstream/Dockerfile:
```diff
@@ -11,6 +12,8 @@ RUN export ARCH=$(case $(arch) in x86_64) echo -n amd64 ;; aarch64) echo -n arm6
     curl -fL -o /usr/local/bin/umoci https://github.com/opencontainers/umoci/releases/download/${UMOCI_VERSION}/umoci.amd64 && \
     chmod a+x /usr/local/bin/umoci && \
     mkdir /project/output && \
+    chmod g+x /entrypoint.sh && \
```
You do not need such step if file bits are already properly set in git repository

samvarankashyap replied:
updated

lslebodn commented on Dockerfiles/midstream/entrypoint.sh:
```diff
@@ -0,0 +1,36 @@
+setup_passwd() {
+    local username="${1:?}"
+    local home="${2:?}"
+    local uid="$3"
```
Looks like we do not pass UID as parameter. 
We can simplify the code and directly use `uid="$(id -u)"`

samvarankashyap replied:
Updated.

lslebodn commented on Dockerfiles/midstream/entrypoint.sh:
```diff
@@ -0,0 +1,36 @@
+    local uid="$3"
+    local gid="${4:-0}"
+    local passwd=/etc/passwd
+    ! whoami 2>/dev/null || return 0
```
We are interested in entry in /etc/passwd. Might be better to check it directly.
```
getent passwd $uid
```

samvarankashyap replied:
Resolved

lslebodn commented on Dockerfiles/midstream/entrypoint.sh:
```diff
@@ -0,0 +1,36 @@
+    local username="${1:?}"
+    [[ -d /workDir ]] && [[ -w /workDir ]] && \
+        mkdir -p "/workDir/home/$username" && \
+        echo "/workDir/home/$username" && \
```
I would rather create it in `/tmp` than in `/workDir`

samvarankashyap replied:
How would it effect the run for changing to /tmp/ ?
I understand it might be the permissions. 
How would this help ?

samvarankashyap replied:
updating meanwhile in subsequent commits

lslebodn replied:
`workDir` is a non-standard directly and we should not rely on it.

lslebodn commented:
@samvarankashyap the script `entrypoint.sh` is written in a way it can be executed always
even for case where entry already exists in /etc/passwd.

But It is assumed users of the image will run it themselves. IMHO we can run it ourselves from the `run_tests.py`
We we can run even tests in gitlab action with passing uid to podman
e.g. `podman run --user 314159:0`

samvarankashyap commented:
@lslebodn Do you mean invoking the entrypoint.sh from run_tests.py 
with appropriate arguments ? 
note: I have borrowed the script from the cpaas with no changes except for calling run_tests.py.

lslebodn commented:
> Do you mean invoking the entrypoint.sh from run_tests.py
Yes

samvarankashyap commented:
@lslebodn 
I tried switiching the logic to run shell script inside the python script run_tests.py 
for reason unknown to me it caused a hungup in both local run and github actions , 
ref: https://github.com/redhat-operator-ecosystem/operator-test-playbooks/pull/152/commits/490b5b426d8f73717c35b3814b10564e93d1427b 

However, When i switched it back to running python script inside the run_tests.py 
it worked absolutely fine. 
ref: https://github.com/redhat-operator-ecosystem/operator-test-playbooks/pull/152/commits/13c100e3895baf5aa312414a8a0fd18b2a423c0d 
I have reverted few changes. However, I made changes to the shell script as requested. 
Please review. 
I will rebase and squash once its re-reviewed 
Thanks. 
Tagging @dirgim @jsztuka for review

dirgim commented on Dockerfiles/midstream/entrypoint.sh:
```diff
@@ -0,0 +1,34 @@
+    local username="${1:?}"
+    local home="${2:?}"
+    local uid="$(id -u)"
+    local gid="${4:-0}"
```
I think it makes sense to change the `gid` to be the 3rd parameter as `uid` isn't tied to a parameter anymore.

jsztuka reviewed (approved):
Looks good, just a small nit that Kruno suggested.

dirgim reviewed (approved):
LGTM

lslebodn commented:
> for reason unknown to me it caused a hungup in both local run and github actions

@samvarankashyap because the cpaas sript was written in a way that would never finish. That was their use-case

Removing `sleep inf` will help
```
diff --git a/Dockerfiles/midstream/fix_etc_passwd.sh b/Dockerfiles/midstream/fix_etc_passwd.sh
index 44b768a..9562285 100644
--- a/Dockerfiles/midstream/fix_etc_passwd.sh
+++ b/Dockerfiles/midstream/fix_etc_passwd.sh
@@ -4,7 +4,6 @@ main() {
     local home
     home=$(mk_home "$username")
     setup_passwd "$username" "$home"
-    exec sleep inf
 }
 mk_home() {
     local username="${1:?}"
```

lslebodn commented:
BTW do not add the script `fix_etc_passwd.sh` into root.
Install it to standard location /usr/bin or /us/local/bin.
It will simplify execution as well. because you will be able to get rid of logic with `fix_etc_passwd_path`
keep it simple and functional :-)

lslebodn commented on Dockerfiles/midstream/run_tests.py:
```diff
@@ -88,6 +88,9 @@ def test_for_extract_and_validate_bundle_image(self):
 if __name__ == '__main__':
     resultcodes = [0, 50, 70, 102, 103]
+    # run fix_etc_passwd.sh to setup random user generated in openshift
+    fix_etc_passwd_path = path.dirname(path.realpath(__file__))
```
unused variable
Otherwise lgtm

samvarankashyap replied:
done

lslebodn commented on Dockerfiles/midstream/fix_etc_passwd.sh:
```diff
@@ -0,0 +1,32 @@
+    local uid="$(id -u)"
+    local gid="${4:-0}"
+    ! whoami 2>/dev/null || return 0
+    if ! [[ $(getent passwd $uid) ]]; then
```
@samvarankashyap I am not sure whether I wrote comment to different line or it was changed on different line.
It was
```
     ! whoami 2>/dev/null || return 0
     if ! [[ -w "$passwd" ]]; then
         echo "There is no permission to add user $username to $passwd"
        return 0
```
And it should be
```
     ! getent passwd $uid 2>/dev/null || return 0
     if ! [[ -w "$passwd" ]]; then
         echo "There is no permission to add user $username to $passwd"
        return 0

```

Eran found the issue
